### PR TITLE
obc: Do not delete user for retain policy if it is owner

### LINKED ISSUE #5510: OBC creation fails when bucket was reused
**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**
OBC creation fails when bucket was reused from old obc with retained policy
**Expected behavior:**
OBC creation should pass

**Additional info**
OBC creation fails because owner of the bucket was missing who was deleted with old obc 

**How to reproduce it (minimal and precise):**

*Create rook cluster with object store running
* Create an OBC using storage-class-retain.yaml and object-bucket-class-retain.yaml
* Delete the OBC
* Now try to create new obc by using existing bucket 
**File(s) to submit**:

* Cluster CR (custom resource), typically called `cluster.yaml`, if necessary
* Operator's logs, if necessary
* Crashing pod(s) logs, if necessary

 To get logs, use `kubectl -n <namespace> logs <pod name>`
When pasting logs, always surround them with backticks or use the `insert code` button from the Github UI.
Read [Github documentation if you need help](https://help.github.com/en/articles/creating-and-highlighting-code-blocks).

**Environment**:
* OS (e.g. from /etc/os-release):
* Kernel (e.g. `uname -a`):
* Cloud provider or hardware configuration:
* Rook version (use `rook version` inside of a Rook Pod):
* Storage backend version (e.g. for ceph do `ceph -v`):
* Kubernetes version (use `kubectl version`):
* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):
* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/rook/master/ceph-toolbox.html)):

### PR DESCRIPTION
**Pull request by thotz**
When obc is deleted, the attached ceph user will be also removed
This will cause obc creation failure for the same bucket since user
was missing. With this patch the user will be retained revoking his permissions

Signed-off-by: Jiffin Tony Thottan <thottanjiffin@gmail.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #5510

**Checklist:**

- [ ] **CommitLint Bot**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/master/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/master/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

[test ceph]

### CODE REVIEW

File: pkg/operator/ceph/object/bucket/s3-handlers.go [hunk 1]
```diff
@@ -176,15 +176,14 @@ const (
 )
 
 // AllowedActions is a lenient default list of actions
-var AllowedActions = []action{
+var OBCAllowedActions = []action{
```
Comment (leseb) [inline pkg/operator/ceph/object/bucket/s3-handlers.go:179]: Why do we need to rename this? Since it's part of the bucket package I think it's clear enough. Also if you really want to rename it you need to change the above comment accordingly.

File: pkg/operator/ceph/object/bucket/s3-handlers.go [hunk 2]
```diff
@@ -176,15 +176,14 @@ const (
...
 	DeleteObject,
 	DeleteObjectVersion,
 	GetBucketAcl,
 	GetBucketCORS,
 	GetBucketLocation,
 	GetBucketLogging,
 	GetBucketNotification,
-	GetBucketPolicy,
```
Comment (leseb) [inline pkg/operator/ceph/object/bucket/s3-handlers.go:187]: How is this related to the main fix? It feels like it should be its own commit unless I'm missing something.

File: pkg/operator/ceph/object/bucket/s3-handlers.go (hunk 1 above)
Comment (thotz) [reply pkg/operator/ceph/object/bucket/s3-handlers.go:179]: > Why do we need to rename this? Since it's part of the bucket package I think it's clear enough. Also if you really want to rename it you need to change the above comment accordingly.

Renaming is not really necessary, try to make it more clear since set of actions used as allow and deny in different circumstances

File: pkg/operator/ceph/object/bucket/s3-handlers.go (hunk 2 above)
Comment (thotz) [reply pkg/operator/ceph/object/bucket/s3-handlers.go:187]: > How is this related to the main fix? It feels like it should be its own commit unless I'm missing something.

Yeah it is not needed, will change in next update

Comment (thotz) [reply pkg/operator/ceph/object/bucket/s3-handlers.go:187]: > > How is this related to the main fix? It feels like it should be its own commit unless I'm missing something.
> 
> Yeah it is not needed, will change in next update

@leseb sorry this change is needed. Otherwise I may need to rewrite actions for the policy in this pr

Comment (leseb) [reply pkg/operator/ceph/object/bucket/s3-handlers.go:187]: It's fine, just explain why it's needed then :)

File: pkg/operator/ceph/object/bucket/s3-handlers.go [hunk 3]
```diff
@@ -176,6 +176,7 @@ const (
 )
 
 // AllowedActions is a lenient default list of actions
+// Removing GetBucketPolicy from the list, refer the commit 43112281
```
Comment (leseb) [inline pkg/operator/ceph/object/bucket/s3-handlers.go:179]: Why do we need this comment? Also the commit ID seems incorrect.

File: pkg/operator/ceph/object/bucket/provisioner.go [hunk 4]
```diff
@@ -254,6 +254,28 @@ func (p Provisioner) Revoke(ob *bktv1alpha1.ObjectBucket) error {
...
+				Actions(AllowedActions...)
+			if policy == nil {
+				policy = NewBucketPolicy(*statement)
+			} else {
+				policy = policy.ModifyBucketPolicy(*statement)
+			}
+			out, err := s3svc.PutBucketPolicy(p.bucketName, *policy)
+			logger.Infof("PutBucketPolicy output: %v", out)
+			if err != nil {
+				return err
```
Comment (leseb) [inline pkg/operator/ceph/object/bucket/provisioner.go:273]: do not just return err but name the error with `errors.Wrap`

Comment (thotz) [reply pkg/operator/ceph/object/bucket/provisioner.go:273]: Done

File: pkg/operator/ceph/object/bucket/s3-handlers.go (hunk 3 above)
Comment (thotz) [reply pkg/operator/ceph/object/bucket/s3-handlers.go:179]: Done

**Comment by thotz** [comment]
> Please include the issue being fixed in your commit message, see: https://rook.io/docs/rook/master/development-flow.html#commit-structure

Done
